data/reusables/actions/jobs/multi-dimension-matrix.md

```diff
@@ -20,3 +20,33 @@ jobs:
         with:
           node-version: {% raw %}${{ matrix.version }}{% endraw %}
 ```
+
+A variable configuration in a matrix can be an `array` of `object`s.
+
+```yaml
+matrix:
+  os:
+    - ubuntu-latest
+    - macos-latest
+  node:
+    - version: 14
+    - version: 20
+      env: NODE_OPTIONS=--openssl-legacy-provider
+```
+
+This matrix produces 4 jobs with corresponding contexts.
+
+```yaml
+- matrix.os: ubuntu-latest
+  matrix.node.version: 14
+  matrix.node.env: ''
```

Comment:
This isn't technically correct — `matrix.node.env` is simply _not set_ for contexts where it's not included as part of the matrix. It's not part of the `matrix.node` object at all. All contexts are not created with the same structure, when matrix expansion is performed on homogeneous objects. Instead, each expanded object has whatever structure is explicitly defined by its set of inputs.

GitHub's workflow syntax expands <code>${{ <var>foo</var> }}</code> to the empty string, for any undefined <var>foo</var>, but that doesn't mean that object actually exists.

Reply:
Which is something that actually requires some care, because for example this workflow:

```yml
jobs:
  build:
    runs-on: ${{ matrix.os.host }}
    strategy:
      matrix:
        os:
          - host: macos-latest
          - host: windows-latest
            shell: 'msys2 {0}'
          - host: ubuntu-latest
        compiler:
          - gcc
          - clang
    defaults:
      run:
        shell: ${{ matrix.os.shell }}
```

...will fail on Ubuntu and macOS, when it tries to set the `run.shell` default value to nothing.

> Error: The template is not valid. foo.yml (Line: Y, Col: X): Unexpected value ''

Reply:
It can be fixed with a fallback, e.g.:

```yml
defaults:
  run:
    shell: ${{ matrix.os.shell || 'bash' }}
```

Reply:
> GitHub's workflow syntax expands ${{ foo }} to the empty string, for any undefined foo, but that doesn't mean that object actually exists.

Where is this documented?

Reply:
Ah, I see ([link](https://docs.github.com/en/actions/learn-github-actions/contexts))

> If you attempt to dereference a nonexistent property, it will evaluate to an empty string.